src/lib/optimizer/chunk_statistics.hpp

```diff
@@ -0,0 +1,171 @@
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+
+#include "storage/base_encoded_column.hpp"
+#include "storage/deprecated_dictionary_column.hpp"
+#include "storage/dictionary_column.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/run_length_column.hpp"
+#include "storage/value_column.hpp"
+
+namespace opossum {
+
+// select how many ranges we want in the filter
+// make this customizable?
+static constexpr uint32_t MAX_RANGES_COUNT = 10;
+
+class BaseFilter : public std::enable_shared_from_this<BaseFilter> {
+ public:
+  virtual ~BaseFilter() = default;
+
+  /**
+   * checks whether the filter is able to determine that the given value 
+   * and predicate condition will not yield any positive results with the values
+   * represented by the filter data.
+   * 
+   * In other words: A scan operation with value and predicate_type on the column/chunk
+   * that this filter was created on would yield zero result rows.
+  */
+  virtual bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const = 0;
+};
+
+class ChunkColumnStatistics {
+ public:
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics(DataType data_type,
+                                                                 std::shared_ptr<BaseColumn> column);
+
+  void add_filter(std::shared_ptr<BaseFilter> filter) { _filters.emplace_back(filter); }
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const {
+    for (const auto& filter : _filters) {
+      if (filter->can_prune(value, predicate_type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+ protected:
+  std::vector<std::shared_ptr<BaseFilter>> _filters;
+};
+
+template <typename T>
+class MinMaxFilter : public BaseFilter {
+ public:
+  explicit MinMaxFilter(T min, T max) : _min(min), _max(max) {}
+  ~MinMaxFilter() override = default;
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    const auto t_value = type_cast<T>(value);
+    // Operators work as follows: value_from_table <operator> t_value
+    // e.g. OpGreaterThan: value_from_table > t_value
+    // thus we can exclude chunk if t_value >= _max since then no value from the table can be greater than t_value
+    switch (predicate_type) {
+      case PredicateCondition::GreaterThan:
+        return t_value >= _max;
+      case PredicateCondition::GreaterThanEquals:
+        return t_value > _max;
+      case PredicateCondition::LessThan:
+        return t_value <= _min;
+      case PredicateCondition::LessThanEquals:
+        return t_value < _min;
+      case PredicateCondition::Equals:
+        return t_value < _min || t_value > _max;
+      default:
+        return false;
+    }
+  }
+
+ protected:
+  const T _min;
+  const T _max;
+};
+
+template <typename T>
+class RangeFilter : public BaseFilter {
```

Comment:
short doc would be nice, name doesn't trivially communicate meaning. What are the ranges? Values contained in the Chunk or those not contained?